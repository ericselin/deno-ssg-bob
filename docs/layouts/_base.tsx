/** @jsx h */

import { Component, getPath, h, readContents } from "../../mod.ts";

const baseCss = getPath(import.meta.url, "_base.css");

export const Base: Component = async (
  { children },
  { needsCss },
) => (
  <html lang="en">
    <head>
      <meta charSet="UTF-8" />
      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <meta http-equiv="X-UA-Compatible" content="ie=edge" />
      <title>bob the static site builder</title>
      <link rel="preconnect" href="https://fonts.gstatic.com" />
      <link
        href="https://fonts.googleapis.com/css2?family=Inter:wght@400;700&family=Source+Code+Pro:wght@400;700;900&display=swap"
        rel="stylesheet"
      />
      <style>
        {await readContents([baseCss, ...needsCss])}
      </style>
    </head>

    <body>
      <header>
        <a href="/">bob</a>
        <ul>
          <li>
            <a href="/getting-started/">Getting started</a>
          </li>
          <li>
            <a href="/scenarios/">Common scenarios</a>
          </li>
        </ul>
      </header>
      {children}
      <footer>
        Static sites should be fast, including updating content
      </footer>
    </body>
  </html>
);
