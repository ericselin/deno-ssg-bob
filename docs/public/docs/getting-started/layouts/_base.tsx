/** @jsx h */

import { Component, h } from "../../../../../mod.ts";

export const Base: Component = (props) => (
  <html>
    <head>
      <title>This is my new site!</title>
    </head>
    <body>
      <header>
        Main header
      </header>
      {props.children}
    </body>
  </html>
);
