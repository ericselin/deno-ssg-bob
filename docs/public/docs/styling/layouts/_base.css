body {
  font-family: serif;
  margin: 1rem auto;
  max-width: 50rem;
}
