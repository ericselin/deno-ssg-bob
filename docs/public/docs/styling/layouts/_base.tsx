/** @jsx h */

import { Component, h, readContents } from "../../../../../mod.ts";

export const Base: Component = async (
  { children },
  { needsCss },
) => (
  <html lang="en">
    <head>
      <title>This is my page</title>
      <style>
        {await readContents(needsCss)}
      </style>
    </head>
    <body>
      {children}
    </body>
  </html>
);

// Tip! You can add CSS files to any component, also this one
Base.needsCss = "_base.css";
