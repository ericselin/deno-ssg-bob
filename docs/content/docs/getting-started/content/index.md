# Awesome new site

This site is based on
[bob](https://github.com/ericselin/deno-ssg-bob).
You should check it out too!

Images work as well!

![Example image](image.jpg)
