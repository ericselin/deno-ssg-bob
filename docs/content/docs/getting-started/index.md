---
title: Getting started
---

Define your layout(s) and write some content. It could not be easier to get going - it "just works".

## Introduction

`bob` uses TSX for templating, which lends itself very well for a modern workflow. Everything is fully typed via Typescript, and as modular as you like.

All of your content lives in the same folder, which mirrors your website structure. E.g. a `blog/post.md` content file gets the full URL `/blog/post/index.html`.

Markdown files in the content folder (`.md`) are redered into HTML, while all other files are copied as-is to the `public` directory.

Example folder structure:

```
site root
â””-- content
|   â””-- index.md      <-- https://example.com/
|   â””-- blog
|       â””-- index.md  <-- https://example.com/blog/
|       â””-- post.md   <-- https://example.com/blog/post/
â””-- layouts
    â””-- _default.tsx  <-- default layout file
    â””-- index.tsx     <-- layout file for index.md
```

Each content file is rendered independently using a layout file based on a layout file lookup order. There is no other magic involved - no base layouts, no partials and nothing like that. However, by importing other layouts and components using stardard ES6 imports, you can create even the most complex layouts. See below!

## 1. Define your layout(s)

In the `layouts` directory, create a `_default.tsx` file with the below contents. This is the default layout file used.

code:layouts/_default.tsx

The above default layout uses the `Base` component, which in our case holds the main HTML skeleton. This is not a requirement by the engine - structuring your layouts is completely up to you! Below is the base component. Note that while `bob` requires you to export layouts as default, components (such as the base) can be consumed in any way. Here we have opted for a named export.

code:layouts/_base.tsx

You can see that we have used `children` in the same way React and React-like libraries do. This is probably a good time to mention the [jsx pragma comment](https://www.typescriptlang.org/tsconfig#jsx) used to inform the Typescript compiler that we want to use the imported `h` function for creating JSX elements.

If you want, create a layout file specific for the root index page.

code:layouts/index.tsx

## 2. Create some content

Creating content is really up to you! But you can of course use the following as a starting point:

code:content/index.md

Please create (or copy over!) more content and make something awesome ðŸ˜Ž.

## Result

This is the resulting index page, which uses the `layouts/index.tsx` layout.

iframe:expected/index.html
