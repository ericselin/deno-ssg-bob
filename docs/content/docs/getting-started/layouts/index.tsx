/** @jsx h */

import { Component, h } from "../../../../../mod.ts";
import { Base } from "./_base.tsx";

const Index: Component = (
  _props,
  { page: { content } }
) => (
  <Base>
    <section>
      <h1>Welcome, welcome</h1>
      <p>
        I wanted to create a better hero
        element, but here we are...
      </p>
    </section>
    <main>
      {content}
    </main>
  </Base>
);

export default Index;
