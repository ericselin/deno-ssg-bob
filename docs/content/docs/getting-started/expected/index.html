<html><head><title>This is my new site!</title></head><body><header>Main header</header><section><h1>Welcome, welcome</h1><p>I wanted to create a better hero element, but here we are...</p></section><main><h1 id="awesome-new-site">Awesome new site</h1>
<p>This site is based on
<a href="https://github.com/ericselin/deno-ssg-bob">bob</a>.
You should check it out too!</p>
<p>Images work as well!</p>
<p><img src="image.jpg" alt="Example image"></p>
</main></body></html>